File: libs/event-bus-data/src/subscriber.rs
//! Consumer side of the data plane bus.
//!
//! Provides at-least-once delivery with **explicit commits**: a record is not
//! considered processed until [`DataMessage::commit`] (or
//! [`DataSubscriber::commit_offsets`]) is called. Auto-commit is off in the
//! default consumer configuration.

use std::sync::Arc;

use async_trait::async_trait;
use rdkafka::consumer::{CommitMode, Consumer, StreamConsumer};
use rdkafka::error::KafkaError;
use rdkafka::message::Message;
use rdkafka::{Offset, TopicPartitionList};
use thiserror::Error;

use crate::config::DataBusConfig;
use crate::headers::OpenLineageHeaders;

#[derive(Debug, Error)]
pub enum SubscribeError {
    #[error("kafka client error: {0}")]
    Kafka(#[from] KafkaError),
    #[error("subscription error: {0}")]
    Subscription(String),
}

#[derive(Debug, Error)]
pub enum CommitError {
    #[error("kafka commit error: {0}")]
    Kafka(#[from] KafkaError),
}

/// One delivered record together with the metadata required to commit it
/// individually.
pub struct DataMessage {
    consumer: Arc<StreamConsumer>,
    topic: String,
    partition: i32,
    offset: i64,
    key: Option<Vec<u8>>,
    payload: Option<Vec<u8>>,
    lineage: Option<OpenLineageHeaders>,
}

impl DataMessage {
    pub fn topic(&self) -> &str {
        &self.topic
    }
    pub fn partition(&self) -> i32 {
        self.partition
    }
    pub fn offset(&self) -> i64 {
        self.offset
    }
    pub fn key(&self) -> Option<&[u8]> {
        self.key.as_deref()
    }
    pub fn payload(&self) -> Option<&[u8]> {
        self.payload.as_deref()
    }
    /// OpenLineage headers extracted from the record, if all required keys
    /// were present and well-formed.
    pub fn lineage(&self) -> Option<&OpenLineageHeaders> {
        self.lineage.as_ref()
    }

    /// Synchronously commit just this record's offset (`offset + 1`).
    pub fn commit(&self) -> Result<(), CommitError> {
        let mut tpl = TopicPartitionList::new();
        tpl.add_partition_offset(&self.topic, self.partition, Offset::Offset(self.offset + 1))
            .map_err(CommitError::Kafka)?;
        self.consumer.commit(&tpl, CommitMode::Sync)?;
        Ok(())
    }
}

impl std::fmt::Debug for DataMessage {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("DataMessage")
            .field("topic", &self.topic)
            .field("partition", &self.partition)
            .field("offset", &self.offset)
            .field("key_len", &self.key.as_ref().map(|k| k.len()))
            .field("payload_len", &self.payload.as_ref().map(|p| p.len()))
            .field("lineage", &self.lineage)
            .finish()
    }
}

/// At-least-once consumer trait.
#[async_trait]
pub trait DataSubscriber: Send + Sync {
    /// Subscribe to one or more topics. Topics must already exist (auto-create
    /// is disabled).
    fn subscribe(&self, topics: &[&str]) -> Result<(), SubscribeError>;

    /// Block until the next record is delivered.
    async fn recv(&self) -> Result<DataMessage, SubscribeError>;

    /// Commit the consumer position for the given message.
    fn commit(&self, message: &DataMessage) -> Result<(), CommitError>;

    /// Commit all currently-stored offsets at once.
    fn commit_offsets(&self) -> Result<(), CommitError>;
}

/// Default `rdkafka`-backed implementation.
#[derive(Clone)]
pub struct KafkaSubscriber {
    consumer: Arc<StreamConsumer>,
}

impl KafkaSubscriber {
    pub fn new(config: &DataBusConfig, group_id: &str) -> Result<Self, SubscribeError> {
        let consumer: StreamConsumer = config.consumer_config(group_id).create()?;
        Ok(Self {
            consumer: Arc::new(consumer),
        })
    }
}

#[async_trait]
impl DataSubscriber for KafkaSubscriber {
    fn subscribe(&self, topics: &[&str]) -> Result<(), SubscribeError> {
        if topics.is_empty() {
            return Err(SubscribeError::Subscription(
                "at least one topic required".into(),
            ));
        }
        self.consumer
            .subscribe(topics)
            .map_err(SubscribeError::Kafka)
    }

    async fn recv(&self) -> Result<DataMessage, SubscribeError> {
        let msg = self.consumer.recv().await?;
        let lineage = msg
            .headers()
            .and_then(OpenLineageHeaders::from_kafka_headers);
        Ok(DataMessage {
            consumer: Arc::clone(&self.consumer),
            topic: msg.topic().to_string(),
            partition: msg.partition(),
            offset: msg.offset(),
            key: msg.key().map(|k| k.to_vec()),
            payload: msg.payload().map(|p| p.to_vec()),
            lineage,
        })
    }

    fn commit(&self, message: &DataMessage) -> Result<(), CommitError> {
        message.commit()
    }

    fn commit_offsets(&self) -> Result<(), CommitError> {
        self.consumer
            .commit_consumer_state(CommitMode::Sync)
            .map_err(CommitError::Kafka)
    }
}

File: libs/event-bus-data/subscriber.go
package databus

import (
	"context"
	"errors"
	"fmt"

	kafka "github.com/segmentio/kafka-go"
)

// DataMessage is one Kafka record exposed to the consumer.
//
// Commit() must be called after the record has been durably processed.
// Auto-commit is off — same contract as the Rust trait.
type DataMessage struct {
	Topic     string
	Partition int
	Offset    int64
	Key       []byte
	Value     []byte
	Headers   []kafka.Header
	Lineage   *OpenLineageHeaders // populated when present in headers

	// internal
	reader *kafka.Reader
	raw    kafka.Message
}

// Commit acknowledges this record's offset to Kafka.
func (m *DataMessage) Commit(ctx context.Context) error {
	if m.reader == nil {
		return errors.New("databus: DataMessage has no associated reader (already committed?)")
	}
	if err := m.reader.CommitMessages(ctx, m.raw); err != nil {
		return fmt.Errorf("commit %s/%d@%d: %w", m.Topic, m.Partition, m.Offset, err)
	}
	return nil
}

// Subscriber is the at-least-once Kafka consumer.
//
// Mirrors the Rust DataSubscriber trait but with batch-commit semantics
// that match segmentio/kafka-go's model:
//
//	for {
//	    msg, err := sub.Poll(ctx)
//	    if err != nil { ... }
//	    batch = append(batch, msg)
//	    if shouldFlush() {
//	        process(batch)
//	        sub.CommitMessages(ctx, batch)
//	    }
//	}
type Subscriber interface {
	Poll(ctx context.Context) (*DataMessage, error)
	// CommitMessages commits the highest offset per topic+partition
	// represented in `msgs`. Pass the slice of every message that
	// has been durably processed since the last commit.
	CommitMessages(ctx context.Context, msgs []*DataMessage) error
	// CommitOffsets is a legacy single-call alternative that commits
	// every offset the underlying reader has seen. Kept for parity
	// with the Rust trait shape; new code should prefer CommitMessages.
	CommitOffsets(ctx context.Context) error
	Close() error
}

// KafkaSubscriber is the segmentio/kafka-go-backed Subscriber.
type KafkaSubscriber struct {
	reader *kafka.Reader
}

// NewKafkaSubscriber wires a Subscriber to one or more topics inside
// the named consumer group. Auto-commit is disabled.
func NewKafkaSubscriber(cfg Config, groupID string, topics []string) (*KafkaSubscriber, error) {
	if len(topics) == 0 {
		return nil, errors.New("databus: at least one topic is required")
	}
	dialer, err := cfg.Principal.dialer(cfg.RequestTimeout)
	if err != nil {
		return nil, fmt.Errorf("kafka dialer: %w", err)
	}
	rcfg := kafka.ReaderConfig{
		Brokers:        cfg.BootstrapServers,
		GroupID:        groupID,
		GroupTopics:    topics,
		Dialer:         dialer,
		IsolationLevel: kafka.ReadCommitted,
		// Auto-commit OFF — caller drives commits explicitly.
		CommitInterval: 0,
	}
	return &KafkaSubscriber{reader: kafka.NewReader(rcfg)}, nil
}

// Poll fetches the next record. Blocks until one arrives or ctx is done.
func (s *KafkaSubscriber) Poll(ctx context.Context) (*DataMessage, error) {
	raw, err := s.reader.FetchMessage(ctx)
	if err != nil {
		return nil, err
	}
	msg := &DataMessage{
		Topic:     raw.Topic,
		Partition: raw.Partition,
		Offset:    raw.Offset,
		Key:       raw.Key,
		Value:     raw.Value,
		Headers:   raw.Headers,
		reader:    s.reader,
		raw:       raw,
	}
	if lineage, ok := OpenLineageHeadersFromKafka(raw.Headers); ok {
		msg.Lineage = &lineage
	}
	return msg, nil
}

// CommitMessages implements Subscriber. Hands the raw messages to the
// underlying segmentio reader, which collapses them to one offset per
// topic+partition before sending the OffsetCommit RPC.
func (s *KafkaSubscriber) CommitMessages(ctx context.Context, msgs []*DataMessage) error {
	if len(msgs) == 0 {
		return nil
	}
	raw := make([]kafka.Message, 0, len(msgs))
	for _, m := range msgs {
		if m == nil {
			continue
		}
		raw = append(raw, m.raw)
	}
	if err := s.reader.CommitMessages(ctx, raw...); err != nil {
		return fmt.Errorf("commit %d messages: %w", len(raw), err)
	}
	return nil
}

// CommitOffsets is a legacy alternative kept for parity with the Rust
// trait. segmentio/kafka-go has no per-reader bulk commit, so this is
// a no-op; new callers should prefer CommitMessages.
func (s *KafkaSubscriber) CommitOffsets(_ context.Context) error { return nil }

// Close releases the underlying Reader.
func (s *KafkaSubscriber) Close() error { return s.reader.Close() }
